apps/web/src/components/Table/ResponsiveTable/ResponsiveTable.tsx

```diff
@@ -0,0 +1,227 @@
+import TablePagination from "./TablePagination";
+import { INITIAL_STATE, SEARCH_PARAM_KEY } from "./constants";
+
+import type {
+  AddLinkProps,
+  DatasetDownload,
+  DatasetPrint,
+  PaginationDef,
+  RowSelectDef,
+  SearchDef,
+  SortDef,
+} from "./types";
+
+interface TableProps<TData> {
+  /** Accessible name for the table */
+  caption: React.ReactNode;
+  /** Data to be displayed within the table */
+  data: TData[];
+  /** Column definitions for `react-table` */
+  columns: ColumnDef<TData>[];
+  /** Column definitions for `react-table` */
+  hiddenColumnIds: string[];
+  /** Determine if any aspect of the table is loading (server side) */
+  isLoading?: boolean;
+  /** Override default null message with a custom one */
+  nullMessage?: NullMessageProps;
+  /** Enable row selection */
+  rowSelect?: RowSelectDef<TData>;
+  /** Enable the search form */
+  search?: SearchDef<TData>;
+  /** Enable sorting */
+  sort?: SortDef;
+  /** Enable pagination */
+  pagination?: PaginationDef;
+  /** Enable printing selected rows (requires rowSelect) */
+  print?: DatasetPrint;
+  /** Enable downloading selected rows and/or all data (requires rowSelect) */
+  download?: DatasetDownload;
+  /** Enable the "add item" button */
+  add?: AddLinkProps;
+}
+
+const ResponsiveTable = <TData extends object>({
+  caption,
+  data,
+  columns,
+  hiddenColumnIds,
+  isLoading,
+  nullMessage,
+  rowSelect,
+  search,
+  sort,
+  download,
+  print,
+  add,
+  pagination,
+}: TableProps<TData>) => {
+  const id = React.useId();
+  const intl = useIntl();
+  const [, setSearchParams] = useSearchParams();
+  const isInternalSearch = search && search.internal;
+  const memoizedData = React.useMemo(() => data, [data]);
+  const memoizedColumns = React.useMemo(() => {
+    if (!rowSelect) return columns;
+    // Inject the selection column if it is enabled
+    return [getRowSelectionColumn(rowSelect.cell, intl), ...columns];
+  }, [columns, intl, rowSelect]);
+  const columnIds = memoizedColumns.map((column) => column.id).filter(notEmpty);
+
+  const [rowSelection, setRowSelection] = React.useState<RowSelectionState>({});
+  const { state, initialState, updaters } = useControlledTableState({
+    columnIds,
+    initialState: {
+      hiddenColumnIds,
+      searchState: search?.initialState,
+      sortState: sort?.initialState,
+      paginationState: pagination?.initialState,
+    },
+  });
+
+  React.useEffect(() => {
+    if (rowSelect?.onRowSelection) {
+      const selectedRows = Object.values(rowSelection)
+        .map((value, index) => {
+          return value ? memoizedData[index] : undefined;
+        })
+        .filter(notEmpty);
+
+      rowSelect.onRowSelection(selectedRows);
+    }
+    // Note: This is probably due to mis-use of useEffect but
+    // adding exhaustive deps causes this to run infinitely
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [rowSelection]);
```

Comment:
I think you can avoid using `useEffect` here by running this code right when `setRowSelection` is called. Maybe by wrapping `setRowSelection`, like this:
```
const mySetRowSelection = (newRowSelectionState) => {
  // call the inner setRowSelection hook method. 
  setRowSelection(newRowSelectionState);
  // other stuff to do
  if (rowSelect?.onRowSelection) {
      const selectedRows = Object.values(newRowSelectionState)
        .map((value, index) => {
          return value ? memoizedData[index] : undefined; 
        })
        .filter(notEmpty);

      rowSelect.onRowSelection(selectedRows);
    }
}
```
Now, if the hook has giving you trouble, something must have an unstable value (memoizedData or rowSelect?) and that might still be a problem, I'm not sure. But avoiding useEffect is recommended where possible.

Reply:
Sweet, that seems to have worked! Turned out to be a little more complicated so I moved it out to it's own hook as well.

[refactor row selection state to avoid useEffect](https://github.com/GCTC-NTGC/gc-digital-talent/pull/7588/commits/f9a070fe195bbf2744e5903b2a75f158fc1b082e)